### PR TITLE
Add support for (Java 7/Java 8) annotation

### PR DESCRIPTION
**Pull request by ChunHin**
This add support for Java 7 or Java 8 annotation to make AsmDataTable more useful.
Useful for modders using Java 7 or Java 8 to get information about annotations.

The gist file for GlobalAnnotationDataTest(With Markdown):
https://gist.github.com/ChunHin/984b987e936d64f9e18c6a9ee4cebdf5

### CODE REVIEW

File: src/main/java/net/minecraftforge/fml/common/discovery/asm/ModMethodVisitor.java [hunk 1]
```diff
@@ -25,4 +33,69 @@ public AnnotationVisitor visitAnnotation(String annotationName, boolean runtimeV
         return new ModAnnotationVisitor(discoverer);
     }
 
+    @Override
+    public AnnotationVisitor visitTypeAnnotation(int typeRef, TypePath typePath, String annotationName, boolean runtimeVisible)
+    {
+        if (typePath != null)
+        {
+            if (typeRef == 335544320)
```
Comment (heldplayer) [inline src/main/java/net/minecraftforge/fml/common/discovery/asm/ModMethodVisitor.java]: Magic value. Where does this come from?

File: src/main/java/net/minecraftforge/fml/common/discovery/asm/ModMethodVisitor.java [hunk 2]
```diff
@@ -25,4 +33,69 @@ public AnnotationVisitor visitAnnotation(String annotationName, boolean runtimeV
...
+            {
+                discoverer.startMethodTypeAnnotation(methodName, typePath.toString(), methodSignatureMap.get(ASMModParser.EnumMethodAdditionalInfo.RETURN_VALUE_TYPE).get(typePath.toString()), annotationName, ASMModParser.EnumMethodAdditionalInfo.RETURN_VALUE_TYPE);
+                return new ModAnnotationVisitor(discoverer);
+            }
+            else if (typeRef == 369098752)
```
Comment (heldplayer) [inline src/main/java/net/minecraftforge/fml/common/discovery/asm/ModMethodVisitor.java]: Same here.

**Comment by ChunHin** [comment]
Going to push a new commit so there are no more magic values.

**Comment by LexManos** [comment]
Why do you need to gather all of this information in the annotation processor?
Also, your description is not really useful.
A more apt description would be something like "Added gathering of annotations on method returns, parameters, and local variables, and class generics"

So ya, what is the use besides academic?

**Pull request closed without merging**